climakitae/threshold_tools.py

```diff
@@ -14,41 +14,162 @@
+def calculate_ess(data, nlags=None):
+    """
+    Function for calculating the effective sample size (ESS) of the provided data.
+
+    Input array is assumed to be timeseries data with potential autocorrelation.
+    """
+    n = len(data)
+    if nlags is None:
+        nlags = n
+    acf = sm.tsa.stattools.acf(data, nlags=nlags, fft=True)
+    sums = 0
+    for k in range(1, len(acf)):
+        sums = sums + (n - k) * acf[k] / n
+
+    return n / (1 + 2 * sums)
 
-    Takes input array and resamples annually by taking the maximum value.
+
+def get_block_maxima(
+    da,
+    extremes_type="max",
+    duration=None,
+    groupby=None,
+    grouped_duration=None,
+    check_ess=True,
+    block_size=1,
+):
+    """
+    Function that converts data into block maximums, defaulting to annual maximums (default block size = 1 year).
+
+    Takes input array and resamples by taking the maximum value over the specified block size.
+
+    Optional arguments `duration`, `groupby`, and `grouped_duration` define the type
+    of event to find the annual maximums of. These correspond to the event
+    types defined in the `get_exceedance_count` function.
 
     Parameters
     ----------
     da: xarray.DataArray
         DataArray from app.retrieve
     extremes_type: str
-        option for max or min (min not implemented yet)
+        option for max or min
         Defaults to max
+    duration: tuple
+        length of extreme event, specified as (4, 'hour')
+    groupby: tuple
+        group over which to look for max occurance, specified as (1, 'day')
+    grouped_duration: tuple
+        length of event after grouping, specified as (5, 'day')
+    check_ess: boolean
+        optional flag specifying whether to check the effective sample size (ESS)
+        within the blocks of data, and throw a warning if the average ESS is too small.
+        can be silenced with check_ess=False.
+    block_size: int
+        block size in years. default is 1 year.
 
     Returns
     -------
     xarray.DataArray
     """
 
-    extremes_types = ["max"]
+    extremes_types = ["max", "min"]
     if extremes_type not in extremes_types:
         raise ValueError(
             "invalid extremes type. expected one of the following: %s" % extremes_types
         )
 
+    # In the simplest case, we use the original data array to take annual
+    # extreme values from
+    da_series = da
+
+    if duration != None:
+        # In this case, user is interested in extreme events lasting at least
+        # as long as the length of `duration`.
+        dur_len, dur_type = duration
+        if dur_type != "hour" or da_series.frequency not in ["1hr", "hourly"]:
+            raise ValueError(
+                "Current specifications not implemented. `duration` options only implemented for `hour` frequency."
+            )
+
+        # First identify the min (max) value for each window of length `duration`
+        if extremes_type == "max":
```

Comment:
Are the extremes_type == max and == min switched? I'm seeing min applied fo max, and max applied for min

Reply:
It's a bit confusing, but pretty sure this is right (but not well documented!! I should add some better explanatory comments). Basically, to get the annual _max_ event lasting for a certain duration (ex: 72 hours), you taking the _min_ value for each 72-hour window of that length, and then to get the annual maximum you take the max of each of these mins...

Reply:
Okay, if I'm following correctly: 

For an annual max event for events of a certain duration (say there are 5 72-hour events in a year), for each of those 5 events, you take the min value of that event which gives you the lowest value above the threshold, and then for the annual statistic you take the max of those "lowest above threshold" values and this gives you... a baseline of a sorts of the most extreme max event?

For annual min, is it then correspondingly the "highest value below the threshold" (so closest to the threshold) and then the min/lowest of each of those "maxes"? 

So in short, we're not looking for the absolute max value in each max event and taking the max of those maxes, which would just be the absolute extreme (and probably easier to calculate!), but rather the max "baseline" of an event above that threshold, and vice versa for mins?